Lib/dataclasses.py

```diff
@@ -1243,6 +1244,34 @@ def _add_slots(cls, is_frozen, weakref_slot):
         if '__setstate__' not in cls_dict:
             cls.__setstate__ = _dataclass_setstate
 
+    # The following is a fix for
+    # https://github.com/python/cpython/issues/111500
+    # The code is copied-and-modified from https://github.com/python-attrs/attrs
+    # All credits for it goes to the `attrs` team and contributors.
+    #
+    # If a method mentions `__class__` or uses the no-arg super(), the
+    # compiler will bake a reference to the class in the method itself
+    # as `method.__closure__`.  Since we replace the class with a
+    # clone, we rewrite these references so it keeps working.
+    for item in cls.__dict__.values():
```

Comment:
Note that this may not work if a method is decorated and surrounded by a wrapper function (e.g., `functools.cache`). Possibly the code could recurse into `.__wrapped__` attributes. That feels risky to me but maybe it's worth doing.

Reply:
> Possibly the code could recurse into `.__wrapped__` attributes. That feels risky to me but maybe it's worth doing.

`inspect.unwrap` goes to some lengths to try to do that safely, and I think `dataclasses` already depends on `inspect`

Reply:
I've added a test case just for this, looks like that the code works without `inspect.unwrap`. Since this is a borrowed code from `attrs`, this gives me additional confidence :)

Reply:
I tried it locally and it passes only if there are other non-decorated methods that use super(). Probably they all share the same cell objects, so if you edit one, the others also work.

On your branch:
```
% cat Lib/test/test_dataclasses/sup.py 
from dataclasses import dataclass
from functools import wraps
def decorator(func):
    @wraps(func)
    def inner(*args, **kwargs):
        return func(*args, **kwargs)
    return inner

@dataclass(slots=True)
class A:
    x: int = 0
    def method(self):
        pass
@dataclass(slots=True)
class B(A):
    @decorator
    def method(self):
        super().method()
B().method()
% ./python.exe Lib/test/test_dataclasses/sup.py
Traceback (most recent call last):
  File "/Users/jelle/py/cpython/Lib/test/test_dataclasses/sup.py", line 19, in <module>
    B().method()
    ~~~~~~~~~~^^
  File "/Users/jelle/py/cpython/Lib/test/test_dataclasses/sup.py", line 6, in inner
    return func(*args, **kwargs)
           ~~~~^^^^^^^^^^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/test/test_dataclasses/sup.py", line 18, in method
    super().method()
    ^^^^^^^^^^^^^^
TypeError: super(type, obj): obj (instance of B) is not an instance or subtype of type (B).
```

Reply:
Oh, this is cool! Adding new tests for this as well.

Reply:
Looks like @carljm already thought of all this last year: https://github.com/python/cpython/issues/90562#issuecomment-1529962640. He also brings up another possible problem with this approach (it breaks the original, non-slotted class). The discussion in that issue seems to conclude that we need a solution in the core instead.

Reply:
I think in the long run dataclasses should have a better approach to slots than class-copying, but it's probably still worth improving class-copying where we can.

This fix would certainly be cleaner if CPython exposed `__classcell__`; then we could easily and reliably find the cell that needs modifying and modify it, without having to rummage through the methods and cross our fingers. I'm not sure if anyone would have objections to that; maybe worth a discourse thread before moving ahead with this version?

Reply:
I case of `__classcell__` it would be impossible to backport this fix. This is my only concern.